fdbserver/include/fdbserver/TagPartitionedLogSystem.actor.h

```diff
@@ -177,18 +179,18 @@ struct TagPartitionedLogSystem final : ILogSystem, ReferenceCounted<TagPartition
 	                                                    LogSystemConfig const& lsConf);
 
 	// Convert TagPartitionedLogSystem to DBCoreState and override input newState as return value
-	void toCoreState(DBCoreState& newState) const final;
+	void toCoreState(DBCoreState& newState) const override final;
 
-	bool remoteStorageRecovered() const final;
+	bool remoteStorageRecovered() const override final;
 
 	// Checks older TLog generations and remove no longer needed generations from the log system.
-	void purgeOldRecoveredGenerations() final;
+	void purgeOldRecoveredGenerations() override final;
 
-	Future<Void> onCoreStateChanged() const final;
+	Future<Void> onCoreStateChanged() const override final;
 
-	void coreStateWritten(DBCoreState const& newState) final;
+	void coreStateWritten(DBCoreState const& newState) override final;
```

Comment:
This `override` is not needed, since `final` is already saying override for the virtual function. So I think we should revert the change.

Reply:
No, the `override` and the `final` are different. `override` means a function *must* be overriding a virtual function in the base class while `final` means the function *cannot* be overrided by its subclass. e.g.

```c++
class Base {
  protected:
    virtual void func_to_be_overridden() = 0;
};

class Derive1 : public Base {
  protected:
    virtual void func_to_be_overridden(int arg) final {}
};

class Derive2 : public Base {
  protected:
    virtual void func_to_be_overridden(int arg) override {}
};
```

clang++ would report
```
test.cpp:13:18: error: 'func_to_be_overridden' marked 'override' but does not override any member functions
    virtual void func_to_be_overridden(int arg) override {}
                 ^
1 error generated.
```
with only one error triggered.

From the standard [1], 11.7.2.4:
```
If a virtual function f in some class B is marked with the virt-specifier final and in a class D derived from B a function D::f overrides B::f, the program is ill-formed.
```
and 11.7.2.5
```
If a virtual function is marked with the virt-specifier override and does not override a member function of a
base class, the program is ill-formed.
```
So basically they are two non-overlapping virt-specifiers.

[1] https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/n4849.pdf

Reply:
`final` also applies to `virtual` function, meaning both `override` and no further subclass override. In your example, the base class uses `virtual void func_to_be_overridden() = 0;` which doesn't take `int arg` as argument, thus a different member function for the `Derive1`.

Reply:
Right, this is the exact example I am trying to show the difference between `final` and `override.`

Note that `final` does not imply that the function *must* override a function, so there is no error popping out when I incorrectly inherit `Base` in `Derive1.` Instead, this problem can only be captured when an actual instance of `Derive1` is constructed. Also another example can be:

```c++
#include <iostream>

struct Base {
    virtual void print() {
        std::cout << "Base" << std::endl;
    }
};

struct Derive : Base {
    // Incorrectly implement the overriding function
    virtual void print(const int v) final {
        std::cout << "Derive " << v << std::endl;
    }
};

int main() {
    auto b = Base();
    auto d = Derive();

    b.print();
    d.Base::print();
    d.print(3);
}
```

Note that `Derive::print` has broken the interface while still getting compiled. This is the reason why `override` is not related to `final.` They are used to capture two different kinds of problems, one from inheriting and the other from being inherited.

Since, in our situation, the `TagPartitionedLogSystem` is trying to inherit from `ILogSystem,` `override` is the more proper decorator for all interface functions. It immediately triggers an error when the virtual function does not share the same interface with the correspondent pure virtual function, rather than complaining function missing during the link stage. I had no idea if the class should be sealed, so I kept the `final,` too.

Reply:
Here is what I propose the style to be used in the code base:

```
class Base {
  protected:
    virtual void func_to_be_overridden() = 0;
};

class Derive1 : public Base {
  protected:
    void func_to_be_overridden(int arg) final {}
};

class Derive2 : public Base {
  protected:
    void func_to_be_overridden(int arg) override {}
};

```
I.e., the derived class do not use "virtual" keyword and "final" or "override" already signify the override of virtual functions.